examples/audio_server/server.cc

```diff
@@ -0,0 +1,132 @@
+#include <memory>
+#include <mutex>
+
+#include "../../blocks/http/api.h"
+#include "../../bricks/dflags/dflags.h"
+#include "../../bricks/sync/waitable_atomic.h"
+
+DEFINE_uint16(port, 8080, "The local port to use.");
+DEFINE_uint32(n, 3, "Max number of audio streams");
+
+const std::string cmd_kill = "kill";
+
+struct SharedState final {
+  uint32_t active_channels = 0;
+  std::map<std::string, std::string> channel_control;
+  bool channel_exists(std::string channel_id) const{
+    return channel_control.find(channel_id) != channel_control.end();
+  }
+};
+
+struct ControlSignal final {
+  std::string apply= "";
+  bool stop = false;
+};
+
+int main(int argc, char** argv) {
+  ParseDFlags(&argc, &argv);
+
+  try {
+    auto& http = HTTP(current::net::AcquireLocalPort(FLAGS_port));
+
+    current::WaitableAtomic<SharedState> safe_state;
+    std::map<std::string, std::thread> threads;
+
+    // Need to join and remove finished sessions
+    auto joiner = std::thread([&threads, &safe_state]{
+      while(true) {
+        std::vector<std::string> to_join;
+        for (auto& pair : threads) {
+          std::string channel_id = pair.first;
+          auto has_channel = safe_state.MutableUse([channel_id](SharedState& state) {
+            return state.channel_exists(channel_id);
+          });
+          if (!has_channel) {
+            to_join.push_back(pair.first);
+            pair.second.join();
+          }
+        }
+        for(auto& to_drop: to_join){
+          threads.erase(to_drop);
+        }
+        std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+      }
+    });
+
+    auto scope = http.Register("/", [&safe_state, &threads](Request r) {
+      std::string channel_id = r.body;
```

Comment:
Maybe URLPathArgs::CountMask::One and use r.url_path_args[0], for localhost:8000/{channel_id}?

If you go this route though, `/kill` might potentially conflict with channel ID. The RESTful standard would be that `/` is just some top-level HATEOAS endpoint (effectively returning the list of endpoints), where you then have `/channel/${id}.

(This way btw, contradictory to my comment below about `/kill`, the canonical deletion would be `curl -X DELETE localhost:8080/channel/{id}`, and `if (r.method == "DELETE")` in the code. Overall, even though I use it a lot myself, `GET` requests are supposed to be legal for proxies to re-try, so for mutating operations such as killing something a non-`GET`, such as `DELETE` or `POST` of `PUT` or `PATCH`, is generally recommended.)

Reply:
in progress